workbench-core/datasets-ui/src/components/CreateInternalDatasetForm.tsx

```diff
@@ -0,0 +1,282 @@
+/*
+ *  Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *  SPDX-License-Identifier: Apache-2.0
+ */
+
+import { ProjectItem } from '@aws/workbench-core-accounts-ui/lib/esm/models/Project';
+import { validateField, FileUpload } from '@aws/workbench-core-swb-common-ui';
+import {
+  Box,
+  SpaceBetween,
+  Form,
+  Header,
+  ExpandableSection,
+  Input,
+  FormField,
+  Button,
+  Container,
+  Textarea,
+  Select,
+  TextContent,
+  Link
+} from '@cloudscape-design/components';
+import React, { useEffect, useState } from 'react';
+import { CreateDatasetForm, CreateDatasetFormValidation } from '../models/Dataset';
+import {
+  datasetDescriptionValidationRules,
+  datasetNameValidationRules,
+  datasetProjectIdValidationRules
+} from '../utils';
+
+interface NewDatasetFormProps {
+  onFormSubmit: (formData: CreateDatasetForm) => Promise<void>;
+  projects: ProjectItem[];
+  areProjectsLoading: boolean;
+}
+
+export const NewDatasetForm = ({
+  onFormSubmit,
+  projects,
+  areProjectsLoading
+}: NewDatasetFormProps): JSX.Element => {
+  const [isSubmitLoading, setIsSubmitLoading] = useState(false);
+  const [disableSubmit, setDisableSubmit] = useState(true);
+  const [error, setError] = useState('');
+  const [formData, setFormData] = useState<CreateDatasetForm>({});
```

Comment:
This might be a "I'm still getting used to how types are used in TypeScript" thing, but it feels weird to pass in an empty object here. That only works because we've declared the fields on `CreateDatasetForm` to all be optional, correct? In practice, all of those fields are required for the request, so I don't think we want to let them be optional. Maybe we could have this be `useState<CreateDatasetForm|null>(null)` and then take away the optionality? What do you think?

Reply:
I agree on comment in general, however I'd like to avoid having `| null`, otherwise we will be doing `?.` everywhere. One way to fix this is to pass empty "correct" object or convert `*Form` into class. And former might be a less of a hustle

Reply:
Updated